h2o-core/src/main/java/hex/grid/FilteredHyperSpaceIterator.java

```diff
@@ -0,0 +1,80 @@
+package hex.grid;
+
+import hex.Model;
+import hex.grid.filter.PermutationFilter;
+
+public class FilteredHyperSpaceIterator<MP extends Model.Parameters> implements HyperSpaceWalker.HyperSpaceIterator<MP> {
+  private HyperSpaceWalker.HyperSpaceIterator<MP> _hyperSpaceIterator;
+
+  private PermutationFilter<MP> _permutationFilter;
+
+  /**
+   * Total number of visited permutations, including those filtered out by {@code _permutationFilter}.
+   */
+  private long _numberOfVisitedPermutations = 0;
+
+  /**
+   * Keeps number of returned to the user permutations as not all visited permutations are considered to be worthy
+   * for evaluation due to a {@code _permutationFilter}. 
+   */
+  private long _numberOfUsedPermutations = 0;
+  private final long _maxHyperSpaceSize;
+
+  public FilteredHyperSpaceIterator(HyperSpaceWalker.HyperSpaceIterator<MP> hyperSpaceIterator, PermutationFilter<MP> permutationFilter, long maxHyperSpaceSize) {
+    _hyperSpaceIterator = hyperSpaceIterator;
+    _permutationFilter = permutationFilter;
+    _maxHyperSpaceSize = maxHyperSpaceSize;
+  }
+
+  @Override
+  public MP nextModelParameters(Model previousModel) {
+    MP nextMPs = _hyperSpaceIterator.nextModelParameters(previousModel);
+    _numberOfVisitedPermutations++;
+    if(! _permutationFilter.test(nextMPs)) {
```

Comment:
I don't like the `test` method name, its not clear whether true means filter-out or filter-in. but I am having a hard time coming up with a better one. Perhaps `if (filter.usePermutation(parms)) {` would be more readable

Reply:
it's not the predicate that decides if you filter in or out, it's how you use it:
- filter in: `stream().filter(x -> predicate.test(x))`
- filter out: `stream().filter(x -> !predicate.test(x))`

Reply:
@honzasterba unfortunately I did a rebase+force push which removed some comments. I had  proposed originally `isPermutationSkipped` ,`skipPermutation`, `skip` names exactly because it has extra meaning that clarifies action:

![image](https://user-images.githubusercontent.com/1162131/69738688-c50b9c00-1136-11ea-8409-50312035b957.png)

@seb-h2o please confirm that you were agains it :)

Reply:
I'd say the confusion may come from the name `permutationFilter`, it's rather `includePermutation.test(nextMPs)`...
 Predicates should always return `true` if they match to avoid multiple negations (either using `!` or negation hidden in the name itself).

Reply:
@seb-h2o  it is weird imp. We have Filter and I don't want to call it like it is and action itself. there might be many methods available on the object. So it is methods name that should provide most of the context

Reply:
Can't disagree more:
```
even = x -> x %2 == 0

I suggest:
even_below_100 = range(100).filter(n -> even.test(n))
odds_below_100 = range(100).filter(n -> !even.test(n))

and you suggest that for each use we need a special method:
even_below_100 = range(100).filter(n -> include_even(n))
odds_below_100 = range(100).filter(n -> exclude_even(n))
```

Reply:
@seb-h2o 
> even = x -> x %2 == 0

here it is not class but instance that you call even. I think that Honza's first comment was about class and its method within it. Your examples are from usage perspective (outside of the class) and it is a cheat kind of as you may call variable/instance whatever you want to explain `.test()` method. It is too much freedom - i will be able to do things like following:

`am_i_a_father_of_my_kids.test()` which does not reflect purpose of the method's functionality.

Reply:
I think the argument with predicate and test() method is valid, I dont think we necessarily need to change the method name, we should just make the line more readable, therefore

```if (usePermutation.test(params) {
}```

is perfectly good solution

Reply:
`predicate.test(x)` always means `is predicate valid for x'.
The problem is not the class, it's how you name your predicate.
`am_i_a_father_of.test(x)` reflects perfectly the purpose: it tells if the predicate `am_I_a_father_of` is valid for x.
coding is language, not logic.

As I said, the confusion can come from `PermutationFilter`, not from the predicate function.... it's because as I tried to encourage before, this class is just a Predicate. It just tells is something is true. And the variable name decides how you use it.
You can have
`filterIn(Predicate includes)`
`filterOut(Predicate excludes)` 
and use the same predicates in both depending on the circumstances.

Reply:
I don't agree with you guys. Next developer will come and will not have this knowledge. It will require him to go and see usages and he/she will be lucky if naming of the instance was chosen carefully. Extra work and error prone. Will rename as 2 > 1

Reply:
I wrote a rules engine in the past where I had hundreds of predicates. I can just tell that if you don't follow the rule above, you stop being able to reason about your code pretty quickly.

Reply:
@seb-h2o let's not justify things by amount of experience. This way we should always listen to our parent :) 
Consequences of your suggestion:
```
@Override
  public void reset() {
    ...
    _includePermutation.reset();
  }
```
Which is what my point was about. I will stop reason pretty quickly at this like for sure.

Reply:
@seb-h2o 

>it's because as I tried to encourage before, this class is just a Predicate. It just tells is something is true. And the variable name decides how you use it.

I'm not saying that I don't see you point.  I just believe that `Predicate.test` is to generic and it does not give you a clue whether you want to filter in or out. You will need some extra context around.

Reply:
It is filtering concept that makes things confusing not Predicate.

Reply:
> It is filtering concept that makes things confusing not Predicate.

Exactly, in most APIs, filter is a verb, not a noun. In electronics/image processing, these are objects, but in programming, these are mainly actions.

> let's not justify things by amount of experience
too bad, that's my favourite argument from authority :) But experience still counts sometimes.

> _includePermutation.reset();
`reset` has an almost universal semantic too, so I don't see the problem here. Besides as it returns `void`, it doesn't need to be interpreted in context. It just resets its argument.

Reply:
@seb-h2o 
> reset has an almost universal semantic too, so I don't see the problem here. Besides as it returns void, it doesn't need to be interpreted in context. It just resets its argument.

Programming is mostly about reading the code. Reseting action/verb does not make sense to me. We should not violate other things to support our ideas. 

My main question: why can't we just rename `test` to something meaningful? Anyway it is not just Predicate as we need Resetable. I think that originally you wanted to have name `test` because you believed that  we can extend `PredicateFilter` and `PermutationFilterFunction` from same interface and that these things are completely composable and interchangeable - but it is not the case.

Reply:
I'm not that religious about `test` or `Predicate`, I could go for `match` otherwise :)
We just have 2 different visions of programming (and maybe of language in general). 
I believe things get their full meaning in context. In programming, this means that a function/class should be as much reusable as possible in different contexts, that's why I chose the most abstract class when possible.
You want to give them an absolute meaning, knowing in advance how they will be used, with the risk of multiplying the classes each time you need a similar but different usage.
Those positions can't be reconciled.
I tried to convince you. I'm failing. No need for me to try harder.

Reply:
@seb-h2o don't make me a bad guy here :) No different visions please. My position is the same -" I chose the most abstract class when possible.". The thing is that it just can't be only Predicate and we need Resetable. For convenience we need alias to this bundle anyway -> no need to use Predicate with its `test` and we can use any name.

Reply:
We probably can have different values/positions which can make things difficult to reconcile but it is not the case here. Here I'm just asking why we can't use another name for `test` if we anyway don't use Predicate class.

Reply:
Well, people have different visions. I'm not saying that ones are always better than others... the proof is that the more this discussion goes, the less we understand each other's concern (I don't have any problem with predicates that can change over time and need to be reset for example). Not trying to make you the bad guy here, really.
At some point, some choice needs to be made, and it's your PR.

Reply:
@seb-h2o all I wanted is to come to a common agreement on the subject. Don't want to just use my vision. But yes, finding common ground takes time.

Reply:
Resetable should be spelled "Resettable"